### PR TITLE
CGE-133 Create entities for chat implementation

### PR DESCRIPTION
**Pull request by drw85**
Created new models to give possibility of implementing chat. 
New logic covered with integration tests.

### CODE REVIEW

File: src/main/java/com/hillel/items_exchange/model/Message.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    @JoinColumn(name = "chat_id", nullable = false)
+    private Chat chat;
+    @ManyToOne
+    @JoinColumn(name = "user_id")
+    private User user;
+    private String text;
+    @OneToMany(mappedBy = "message", cascade = CascadeType.ALL, orphanRemoval = true)
+    List<Attachment> attachments;
+    private LocalDateTime created;
+    private LocalDateTime updated;
```
Comment (Jack11M) [inline src/main/java/com/hillel/items_exchange/model/Message.java:31]: Why don't you extend BaseEntity and use fields: id, created, updated from it?

File: src/main/java/com/hillel/items_exchange/model/Chat.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    @Column(name = "hash", nullable = false, unique = true)
+    private String hash;
+    @ManyToOne(cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "advertisement_id", nullable = false)
+    private Advertisement advertisement;
+    @ManyToMany(mappedBy = "chats")
+    private List<User> users;
+    @OneToMany(mappedBy = "chat", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<Message> messages;
+}
```
Comment (Jack11M) [inline src/main/java/com/hillel/items_exchange/model/Chat.java:29]: GitHub requires a blank line at the end of each file.

File: src/main/java/com/hillel/items_exchange/model/Message.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    @Enumerated(EnumType.STRING)
+    private MessageStatus status;
+
+    @PrePersist
+    private void onPrePersist() {
+        created = LocalDateTime.now();
+        updated = LocalDateTime.now();
+        status = MessageStatus.NEW;
```
Comment (Jack11M) [inline src/main/java/com/hillel/items_exchange/model/Message.java:39]: Great approach!

File: src/main/java/com/hillel/items_exchange/model/Message.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @PreUpdate
+    private void onPreUpdate() {
+        updated = LocalDateTime.now();
+    }
+}
```
Comment (Jack11M) [inline src/main/java/com/hillel/items_exchange/model/Message.java:46]: GitHub requires a blank line at the end of each file.

File: src/main/resources/application.properties [hunk 5]
```diff
@@ -43,4 +43,4 @@ app.image.thumbnail.edge.px=300
 logging.level.org.springframework.web=DEBUG
 logging.pattern.file= "%d{yyyy-MM-dd HH:mm:ss} [%thread] %-5level %logger{36} - %msg%n"
 logging.file.name=LogFile.log
-logging.file.path=${logs-path}
\ No newline at end of file
+#logging.file.path=${logs-path}
```
Comment (Jack11M) [inline src/main/resources/application.properties:46]: GitHub requires a blank line at the end of each file.

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.Collections;
+import java.util.List;
+
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+@DBRider
+@DBUnit(cacheConnection = false)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+public class ChatEntitiesIntegrationTest {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:29]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    AttachmentRepository attachmentRepository;
+
+    private final String CHAT_HASH = "TestHash";
+    private final String FIRST_USER_USERNAME = "test";
+    private final String SECOND_USER_USERNAME = "test2";
+
+    @Test
+    @DataSet("chat/init_db.yml")
+    @ExpectedDataSet(value = {"chat/init_db.yml", "chat/add_new_chat_to_user.yml"})
+    public void shouldAddNewChatToUser() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:48]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                new ArrayList<>(List.of(user)), Collections.emptyList());
+        user.getChats().add(chat);
+        userRepository.saveAndFlush(user);
+    }
+
+    @Test
+    @DataSet(value = {"chat/init_db.yml", "chat/add_new_chat_to_user.yml"})
+    @ExpectedDataSet(value = {"chat/init_db.yml", "chat/add_user_to_existing_chat.yml"},
+            ignoreCols = "ID")
+    public void shouldAddAnotherUserToChat() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:60]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Test
+    @DataSet(value = {"chat/init_db.yml", "chat/add_new_chat_to_user.yml"})
+    @ExpectedDataSet(value = {"chat/init_db.yml",
+            "chat/add_new_chat_to_user.yml",
+            "chat/add_new_message_without_attachments_to_chat.yml"},
+            ignoreCols = {"created", "updated", "id"})
+    @ExportDataSet(format = DataSetFormat.YML, outputName = "target/exported/testing.yml")
+    public void shouldAddNewMessageWithNoAttachmentsToUsersChat() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:74]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Test
+    @DataSet(value = {"chat/init_db.yml", "chat/add_user_to_existing_chat.yml"})
+    @ExpectedDataSet(value = {"chat/init_db.yml",
+            "chat/add_user_to_existing_chat.yml",
+            "chat/add_message_with_attachments_for_every_user_in_chat.yml"},
+            ignoreCols = {"created", "updated", "file_content"})
+    @ExportDataSet(format = DataSetFormat.YML, outputName = "target/exported/testing.yml")
+    public void shouldAddNewMessageWithTwoAttachmentsForAllUsersOfChat() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:89]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        });
+    }
+
+    @Test
+    @DataSet(value = {"chat/init_db.yml",
+            "chat/add_new_chat_to_user.yml",
+            "chat/add_new_message_without_attachments_to_chat.yml"})
+    @ExpectedDataSet("chat/remove_user_and_chats.yml")
+    @ExportDataSet(format = DataSetFormat.YML, outputName = "target/exported/testing.yml")
+    public void shouldRemoveUserWithAllOfHisChatsAndMessagesWithAttachments() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:108]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Test
+    @DataSet(value = {"chat/init_db.yml",
+            "chat/add_user_to_existing_chat.yml",
+            "chat/add_message_with_attachments_for_every_user_in_chat.yml"})
+    @ExpectedDataSet(value = "chat/remove_user_from_chat_with_more_than_one_user.yml",
+            ignoreCols = {"created", "updated", "file_content"})
+    @ExportDataSet(format = DataSetFormat.YML, outputName = "target/exported/testing.yml")
+    public void shouldRemoveUserWithHisMessagesAndAttachmentsForThemButLeaveChat() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:121]: The public modifier is redundant here

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        final User existingUser = getExistingUserByUsername(SECOND_USER_USERNAME);
+        existingUser.getChats().remove(0);
+        userRepository.delete(existingUser);
+        userRepository.flush();
+    }
+
+    @Test
+    @DataSet(value = {"chat/init_db.yml"})
+    @ExpectedDataSet("chat/add_user_to_blacklist.yml")
+    public void shouldAddUserToBlacklist() {
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:131]: The public modifier is redundant here

File: src/test/resources/datasets/chat/add_user_to_existing_chat.yml [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+    hash: 'TestHash'
+    ADVERTISEMENT_ID: 1
+
+USER_CHAT:
+  - ID: 1
+    USER_ID: 2
+    CHAT_ID: 1
+  - ID: 2
+    USER_ID: 1
+    CHAT_ID: 1
```
Comment (Jack11M) [inline src/test/resources/datasets/chat/add_user_to_existing_chat.yml:12]: GitHub requires a blank line at the end of each file.

File: src/test/resources/datasets/chat/add_user_to_blacklist.yml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+BLACK_LIST:
+  - ID: 1
+    BLOCKER_ID: 1
+    BLOCKED_ID: 2
```
Comment (Jack11M) [inline src/test/resources/datasets/chat/add_user_to_blacklist.yml:4]: GitHub requires a blank line at the end of each file.

File: src/test/resources/datasets/chat/add_new_message_without_attachments_to_chat.yml [hunk 16]
```diff
@@ -0,0 +1,8 @@
+MESSAGE:
+  - ID: 1
+    CHAT_ID: 1
+    USER_ID: 2
+    TEXT: "Test message text of user test2"
+    STATUS: "NEW"
+    CREATED: "2020-10-29 14:02:58.482885"
+    UPDATED: "2020-10-29 14:02:58.504797"
```
Comment (Jack11M) [inline src/test/resources/datasets/chat/add_new_message_without_attachments_to_chat.yml:8]: GitHub requires a blank line at the end of each file.

File: src/test/resources/datasets/chat/add_new_chat_to_user.yml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+CHAT:
+  - ID: 1
+    hash: 'TestHash'
+    ADVERTISEMENT_ID: 1
+
+USER_CHAT:
+  - ID: 1
+    USER_ID: 2
+    CHAT_ID: 1
```
Comment (Jack11M) [inline src/test/resources/datasets/chat/add_new_chat_to_user.yml:9]: GitHub requires a blank line at the end of each file.

File: src/test/resources/datasets/chat/add_new_attachment_to_message.yml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+ATTACHMENT:
+  - ID: 1
+    MESSAGE_ID: 1
+    TYPE: "jpg"
+    FILE_CONTENT: "ZmlsZV9jb250ZW50"
+  - ID: 2
+    MESSAGE_ID: 1
+    TYPE: "png"
+    FILE_CONTENT: "ZmlsZV9jb250ZW50Mg=="
```
Comment (Jack11M) [inline src/test/resources/datasets/chat/add_new_attachment_to_message.yml:9]: GitHub requires a blank line at the end of each file.

File: src/test/resources/datasets/chat/add_message_with_attachments_for_every_user_in_chat.yml [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    TYPE: "jpg"
+    FILE_CONTENT: "ZGVm"
+  - ID: 3
+    MESSAGE_ID: 2
+    TYPE: "png"
+    FILE_CONTENT: "YWJj"
+  - ID: 4
+    MESSAGE_ID: 2
+    TYPE: "jpg"
+    FILE_CONTENT: "ZGVm"
```
Comment (Jack11M) [inline src/test/resources/datasets/chat/add_message_with_attachments_for_every_user_in_chat.yml:33]: GitHub requires a blank line at the end of each file.

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+    private Message createMessageWithoutAttachments(User user, Chat chat) {
+        return new Message(0L, chat, user, "Test message text of user " + user.getUsername(),
+                Collections.emptyList(), null, null, MessageStatus.NEW);
+    }
+
+    private Attachment createAttachment(Message message, String type) {
+        return new Attachment(0L, message, type, "file_content".getBytes());
+    }
+}
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:158]: GitHub requires a blank line at the end of each file.

File: src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+        final User user1 = getExistingUserByUsername(FIRST_USER_USERNAME);
+        final User user2 = getExistingUserByUsername(SECOND_USER_USERNAME);
+        user1.getBlacklistedUsers().add(user2);
+        userRepository.saveAndFlush(user1);
+    }
+
+    private User getExistingUserByUsername(String username) {
+        return userRepository.findByUsername(username).orElseThrow(EntityNotFoundException::new);
+    }
+
```
Comment (Jack11M) [inline src/test/java/com/hillel/items_exchange/chat/ChatEntitiesIntegrationTest.java:141]: As far as I know, using helper methods in Test classes is bad practice, you can port them to TestUtil instead.

**Review by Jack11M — COMMENTED** [review]
Lots of work!
Please, reply to my comments.

Also when I run all tests, the updateAdvertisement_shouldUpdateExistedAdvertisement() in AdvertisementControllerIntegrationTest fails (please, check it, maybe this is only my problem).

Error message:  junit.framework.ComparisonFailure: row count (table=IMAGE) expected:<[1]> but was:<[2]>

**Review by rpkyrych — CHANGES_REQUESTED** [review]
Nice job but few things needs to be updated.

File: src/main/java/com/hillel/items_exchange/model/User.java [hunk 22]
```diff
@@ -53,4 +53,19 @@
 
     @OneToMany(mappedBy = "user", cascade = CascadeType.ALL, orphanRemoval = true)
     private Set<Child> children;
+
+    @ManyToMany(cascade = {CascadeType.ALL})
```
Comment (lubchik-14) [inline src/main/java/com/hillel/items_exchange/model/User.java:57]: I am confused about behavior that we can observe deleting a user. CascadeType.ALL propagates deleting child entity this behavior is not every needed us. Maybe should not use this type here?

**Review by lubchik-14 — CHANGES_REQUESTED** [review]
So, it's a super job, but I left some comments. 
I also have failed tests AdvertisementControllerIntegrationTest.updateAdvertisement_shouldUpdateExistedAdvertisement somehow.

**Pull request closed without merging**